cmd/kube-controller-manager/app/testing/testserver.go

```diff
@@ -75,27 +79,31 @@ func StartTestServer(ctx context.Context, customFlags []string) (result TestServ
 				logger.Error(err, "Failed to shutdown test server cleanly")
 			}
 		}
-		if len(result.TmpDir) != 0 {
-			os.RemoveAll(result.TmpDir)
-		}
 	}
 	defer func() {
 		if result.TearDownFn == nil {
 			tearDown()
 		}
 	}()
 
-	result.TmpDir, err = os.MkdirTemp("", "kube-controller-manager")
-	if err != nil {
-		return result, fmt.Errorf("failed to create temp dir: %v", err)
-	}
+	result.TmpDir = t.TempDir()
 
 	fs := pflag.NewFlagSet("test", pflag.PanicOnError)
 
 	s, err := options.NewKubeControllerManagerOptions()
 	if err != nil {
 		return TestServer{}, err
 	}
+	// set up new instance of ComponentGlobalsRegistry instead of using the DefaultComponentGlobalsRegistry to avoid contention in parallel tests.
+	featureGate := utilfeature.DefaultMutableFeatureGate.DeepCopy()
```

Comment:
just so I understand, this is copying the state of the global bits at kcm testing server init time, then applying the flags, then detecting if the local copies are different from the global copies and trying to propagate the changes to the global copies using `Set___DuringTest`?

I think that means any calls to `featuregatetesting.SetFeatureGateDuringTest` *before* starting the test servers will get picked up in the initial deepcopies, which is fine.

we don't use this testing package in that many places, I'm trying to figure out how TestClusterTrustBundlesPublisherController will behave. It first starts a test API server, then a test KCM server, and passes `--feature-gates=ClusterTrustBundle=true` to both of them